README.md
# Easy Mask

Easy way to apply a mask to Flutter's TextFields and on Strings

To use it, you only need to pass EasyMask on TextField's parameter inputFormatters.

The Easy Mask cares about user's cursor position, to make fluid it's usability.

Supports Multi Mask based on masks length.

Supports Placeholders.

![Sample](https://raw.githubusercontent.com/danilocoppi/flutter-textfield-mask/main/img/sample.gif)

# Usage

### **Simplest Sample**

`import 'package:easy_mask/easy_mask.dart';`

Then instantiate `TextInputMask` passing at least a String `mask` parameter.

``` dart
import 'package:easy_mask/easy_mask.dart';
...
  TextField(
    inputFormatters: [ TextInputMask(mask: '99? (99) 999 99-99') ],
  ),
  TextField(
    inputFormatters: [ TextInputMask(mask: '999.999.999-99', reverse:true ) ],
  )
...
```

Formatting a String

``` dart
import 'package:easy_mask/easy_mask.dart';
...
  String text = '432516565';
  MagicMask mask = MagicMask.buildMask('\\+99 (99) 99999-9999');
  String formattedString = mask.getMaskedString(text);
...
```

### **Multi Mask Sample**

`import 'package:easy_mask/easy_mask.dart';`

Then instantiate `TextInputMask` passing at least a String `mask` parameter.

``` dart
import 'package:easy_mask/easy_mask.dart';
...
  TextField(
    inputFormatters: [ TextInputMask(mask: ['999.999.999-99', '99.999.999/9999-99'] ],
  ),
  TextField(
    inputFormatters: [ TextInputMask(mask: ['(99) 9999 9999', '(99) 99999 9999'], reverse:true ) ],
  )
...
```

### **PlaceHolder Sample**

`import 'package:easy_mask/easy_mask.dart';`

Then instantiate `TextInputMask` passing at least a String `mask` parameter.

``` dart
import 'package:easy_mask/easy_mask.dart';
...
  TextField(
    inputFormatters: [ TextInputMask(mask: '999.999.999-99', placeholder: '_', maxPlaceHolders: 11) ],
  ),
...
```

### **Pretty Currency Sample**

`import 'package:easy_mask/easy_mask.dart';`

Then instantiate `TextInputMask` passing at least a String `mask` parameter.

``` dart
import 'package:easy_mask/easy_mask.dart';
...
  TextField(
    inputFormatters: [ TextInputMask(mask: '\$! !9+,999.99', placeholder: '0', maxPlaceHolders: 3, reverse: true)],
  ),
...
```

### Possible parameters

* **mask** can be a String or an List<String> with the wanted masks pattern.
* **reverse** is a boolean that indicates if the user will type from right to left. Used normally on currency TextFields.
* **maxLength** is an Integer that limits the maximum size of returned masked text.  
* **placeholder** is a String character to be used as placeholder on untyped characters. Must define maxPlaceHolders
* **maxPlaceHolders** an integer to map how many possible places it would be placed. Typed characters consumes a position from this counter.

### Mask Patterns Characters

 **9** - is used to allow a number from 0-9

 **A** - is used to allow a letter from a-z or A-Z

 **N** - is used to allow a number or letter from 0-9, a-z or A-Z

 **X** - is used to allow any character

#### *Those tokens 9,A,N and X can be followed by one following modifier*

 **?** - indicates that is optional

 **\+** - indicates that must have at least 1 or more repetitions

 **\*** - indicates that can have 0 or more repetitions

 **\\** - is used as scape

#### *Any character that is interpreted as letter to be placed, can be followed by modifier **

 **\!** - Used to force print it, when it has at least 1 character typed on TextField.

Any other letters will be displayed on masking.

#### *Examples of String masks*

(card number) 999 999 999 999

(us cellphone) \\\+1 (999) 999 99 99

(currency) $! !9+,999.99

(version) 99?9?.99?9?.99?9?

(RG brazilian document) 99.999.999-N

CHANGELOG.md
## [2.0.2]

* Fixed regex for letter input to reject "_" (underscore)

## 2.0.1

* Fixed _overflow has not been initialized

## 2.0.0

* Migrate to null safety

## 1.1.5

* `Fix problems with stuck characters` - Fix the problem with amount inverted mask with placeholders that wasn't working properly

## 1.1.4

* `Fix Method Signature` - Fix wrong method signature, that was throwing an exception when using list of masks (Fixed by DeveJonatas)

## 1.1.3

* `Fix Readme` - Fixed Readme problem

## 1.1.2

* `String Format` - Added easy way to format strings using it.

## 1.1.1

* `Dartfmt` - Applying dart formats.

## 1.1.0

* `PlaceHolder` - Added support to placeholders.
* `MultiMask` - Support multi masking, that changes based on length typed.
* `Feature !` - Added the maker ! for force a char to be displayed on formatted text.

## 1.0.2

* `ReadMe` - Improved the Readme.

## 1.0.1

* `Altered ReadMe` - Added GIF sample result.

## 1.0.0

* `First Stable Version` - Ready to be used.

## 0.0.1

* `Initial Version` - Developing.


test/main_test.dart
import 'package:easy_mask/src/magic_mask.dart';
import 'package:flutter_test/flutter_test.dart';

void main() async {
  test('Brazilian Cellphone', () {
    String baseMask = '\\+99? (99) 99999 - 9999?';
    String baseTest = '5519981234567';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 3, false, -1, '', 0);
    print(result);
    expect(result['text'], '+55 (19) 98123 - 4567');
    expect(result['selectionBase'], 6);
  });

  test('Brazilian Personal Document', () {
    String baseMask = '999.999.999-99';
    String baseTest = '123123HUM12344';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 4, false, -1, '', 0);
    print(result);
    expect(result['text'], '123.123.123-44');
    expect(result['selectionBase'], 5);
  });

  test('USA Celphone', () {
    String baseMask = '\\+1 (999) 999 99 99';
    String baseTest = '446667AAAAAA8899';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 4, false, -1, '', 0);
    print(result);
    expect(result['text'], '+1 (446) 667 88 99');
    expect(result['selectionBase'], 10);
  });

  test('Credit Card', () {
    String baseMask = '9999 9999 9999 9999';
    String baseTest = '1234555@#566667878';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 4, false, -1, '', 0);
    print(result);
    expect(result['text'], '1234 5555 6666 7878');
    expect(result['selectionBase'], 4);
  });

  test('Reversed Money without thousand separator comma', () {
    String baseMask = '\$! !9+.99';
    String baseTest = '1025065';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 7, true, -1, '', 0);
    print(result);
    expect(result['text'], '\$ 10250.65');
    expect(result['selectionBase'], 10);
  });

  test('Reversed Money with thousand separator comma', () {
    String baseMask = '\$! !9+,999.99';
    String baseTest = '1025065A';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 7, true, -1, '', 0);
    print(result);
    expect(result['text'], '\$ 10,250.65');
    expect(result['selectionBase'], 11);
  });

  test('Custom Masks', () {
    String baseMask = '99 AA 999';
    String baseTest = '33xyz.@999';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 7, false, -1, '', 0);
    print(result);
    expect(result['text'], '33 xy 999');
    expect(result['selectionBase'], 5);
  });

  test('Escape Masks', () {
    String baseMask = '\\9! \\A! AAAA';
    String baseTest = 'Test';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 0, false, -1, '', 0);
    print(result);
    expect(result['text'], '9 A Test');
    expect(result['selectionBase'], 0);
  });

  test('Escape Masks(Time Case)', () {
    String baseMask = '99:99 !\\A!M!';
    String baseTest = '12:30';
    MagicMask mm = MagicMask();
    mm.buildMaskTokens(baseMask);
    var result = mm.executeMasking(baseTest, 0, false, -1, '', 0);
    print(result);
    expect(result['text'], '12:30 AM');
    expect(result['selectionBase'], 0);
  });

  test('Simple text', () {
    String text = '432516565';
    MagicMask mask = MagicMask.buildMask('\\+99 (99) 99999-9999');
    var res = mask.getMaskedString(text);
    expect(res, '+43 (25) 16565');
  });
}


lib/src/magic_mask.dart
import 'dart:math';

/// This class is used to make all hardwork of making and controll the users cursor.
/// It should be used by TextInputMask, so you dont need to worry about it.
/// If you want you can use it as String masker.
class MagicMask {
  static const String _type = 'type';
  static const String _value = 'value';

  static const String _fixChar = 'fixedChar';
  static const String _forcedChar = 'forcedChar';
  static const String _token = 'token';
  static const String _tokenOpt = 'optionalToken';
  static const String _multiple = 'multiple';
  static const String _multipleOpt = 'multiple';

  late bool _reverse;
  bool _overflow = false;
  late int _charIndex;
  late int _tagIndex;
  late int _step;
  late int _charDeslocation;
  late int _cursorPosition;
  late String _placeholder;
  late int _maxPlaceHolderCharacters;
  late String _maskedText;
  late String _extraChar;
  late int _typedCharacter;

  List<Map<String, String>> _tags = [];
  List<List<Map<String, String>>> _allTags = [];
  int _curTag = 0;

  MagicMask();

  MagicMask.buildMask(dynamic mask) {
    if (mask != null) this.buildMaskTokens(mask);
  }

  String? _lastMaskType() => _tags.last[_type];

  /// the BuildMaskTokens will transform the String pattern in tokens to be used as formatter.
  /// The [mask] should a String following the pattern:
  ///
  /// 9 - is used to allow a number from 0-9
  ///
  /// A - is used to allow a letter from a-z or A-Z
  ///
  /// N - is used to allow a number or letter from 0-9, a-z or A-Z
  ///
  /// X - is used to allow any character
  ///
  /// **Those tokens 9,A,N and X can be followed by one following modifier**
  ///
  /// \? - indicates that is optional
  ///
  /// \+ - indicates that must have at least 1 or more repetitions (use as last character)
  ///
  /// \* - indicates that can have 0 or more repetitions (use as last character)
  ///
  /// \ - is used as scape
  ///
  ///
  /// ** Any character that is interpreted as letter to be placed, can be followed by modifier **
  ///
  /// \! - Used to force print it, when it has at least 1 letter
  void buildMaskTokens(dynamic masks) {
    List<String> maskList = [];
    if (masks is String) {
      maskList.add(masks);
    } else if (masks is List<String>) {
      maskList = masks;
    } else {
      throw Exception('Unknown mask type');
    }
    _curTag = 0;
    for (var mask in maskList) {
      _allTags.add([]);
      _tags = _allTags[_curTag];
      _processMask(mask);
      _curTag += 1;
    }

    _allTags.sort((maskA, maskB) => maskA.length > maskB.length
        ? 1
        : maskA.length == maskB.length
            ? 0
            : -1);
  }

  void _processMask(String mask) {
    for (var i = 0; i < mask.length; i++) {
      String currentChar = mask[i];
      if (currentChar == '\\') {
        _tags.add({_type: _fixChar, _value: mask[i + 1]});
      } else if (currentChar == '*') {
        if (_lastMaskType() == _token) _tags.last[_type] = _multipleOpt;
      } else if (currentChar == '+') {
        if (_lastMaskType() == _token) _tags.last[_type] = _multiple;
      } else if (currentChar == '?') {
        if (_lastMaskType() == _token) _tags.last[_type] = _tokenOpt;
      } else if (currentChar == '!') {
        if (_lastMaskType() == _fixChar) _tags.last[_type] = _forcedChar;
      } else if (currentChar == '9') {
        _tags.add({_type: _token, _value: '\\d'});
      } else if (currentChar == 'A') {
        _tags.add({_type: _token, _value: '[a-zA-Z]'});
      } else if (currentChar == 'N') {
        _tags.add({_type: _token, _value: '[a-zA-Z0-9]'});
      } else if (currentChar == 'X') {
        _tags.add({_type: _token, _value: '.'});
      } else {
        _tags.add({_type: _fixChar, _value: currentChar});
      }
    }
  }

  /// [text] is the string to be formatted
  /// It returns a formatted String.
  String getMaskedString(String text) =>
      getAdvancedMaskedString(text, -1, '', 0);

  /// [text] is the string to be formatted
  /// [maxLenght] is used to limit the maximum returned text. Set it as -1 to not limitate.
  /// [placeholder] String character to be applyed as placeholder
  /// [maxPlaceHolderCharacters] Numbers of times the placeholder could be counted. A typed character consumes a count.
  ///
  /// It returns a formatted String.
  String getAdvancedMaskedString(String text, int maxLenght, String placeholder,
      int maxPlaceHolderCharacters) {
    return executeMasking(text, 0, false, maxLenght, placeholder,
        maxPlaceHolderCharacters)['text'];
  }

  /// [text] is the mask to be formatter on mask.
  /// [cursorPosition] means the cursor position before masking.
  /// [reverse] is used to define the diretion mask will be applyed.
  /// [maxLenght] is used to limit the maximum returned text. Set it as -1 to not limitate.
  /// [placeholder] String character to be applyed as placeholder
  /// [maxPlaceHolderCharacters] Numbers of times the placeholder could be counted. A typed character consumes a count.
  ///
  /// It Return a JSON format to be used to create a TextEditingValue. Its format is:
  /// ```
  /// {
  ///   "text": formattedText,
  ///   "selectionBase": newCursorPosition,
  ///   "selectionExtent": newCursorPosition
  /// }
  /// ```
  ///
  Map<String, dynamic> executeMasking(
      String? text,
      int cursorPosition,
      bool reverse,
      int maxLenght,
      String placeholder,
      int maxPlaceHolderCharacters) {
    if (text == null || text.isEmpty || _tags.length == 0)
      return _buildResultJson('', 0, maxLenght);
    _reverse = reverse;
    _step = _reverse ? -1 : 1;
    _placeholder = placeholder;
    _maxPlaceHolderCharacters = maxPlaceHolderCharacters;
    List<Map<String, dynamic>> results = [];
    for (var i = 0; i < _allTags.length; i++) {
      _tags = _allTags[i];
      _cursorPosition = cursorPosition;
      _charDeslocation = 0;
      _typedCharacter = 0;
      _maskedText = '';
      _extraChar = '';
      _overflow = false;
      _tagIndex = _reverse ? _tags.length - 1 : 0;
      for (Map<String, String> tag in _tags) tag['readed'] = '';

      String cleared = clearMask(text);
      cleared = _clearPlaceHolder(cleared);
      if (cleared.isEmpty) return _buildResultJson('', 0, maxLenght);
      Map<String, dynamic> res = _proccessMask(cleared, maxLenght);
      if (res['overflow'] == false) return res;
      results.add(res);
    }
    return results.last;
  }

  String _clearPlaceHolder(String text) {
    String cleared = text;
    int index = _reverse ? 0 : text.length - 1;
    while (index >= 0 && index < text.length && text[index] == _placeholder) {
      cleared =
          _reverse ? text.substring(1) : text.substring(0, text.length - 1);
      index -= _step;
    }
    return cleared;
  }

  /// [text] is the text with mask that should be cleared.
  ///
  /// It Return a String without the mask:
  ///
  String clearMask(String text) {
    String cleared = text;
    int tagIndex = _reverse ? _tags.length - 1 : 0;
    while (tagIndex >= 0 && tagIndex < _tags.length) {
      var tag = _tags[tagIndex];
      if (tag[_type] == _forcedChar || tag[_type] == _fixChar) {
        int pos = _reverse
            ? cleared.lastIndexOf(tag[_value]!)
            : cleared.indexOf(tag[_value]!);
        if (pos != -1) {
          if (pos < _cursorPosition) {
            _cursorPosition -= 1;
          }
          cleared = '${cleared.substring(0, pos)}${cleared.substring(pos + 1)}';
        } else {
          break;
        }
      }
      tagIndex += _step;
    }
    return cleared;
  }

  Map<String, dynamic> _proccessMask(String text, int maxLenght) {
    _charIndex = _reverse ? text.length - 1 : 0;
    String currentChar = text[_charIndex];
    while (currentChar.isNotEmpty) {
      _applyTagMask(currentChar, false);
      _charIndex += _step;
      if (_charIndex < 0 || _charIndex >= text.length) break;
      currentChar = text[_charIndex];
    }

    int placeHolderCounter = _maxPlaceHolderCharacters - _typedCharacter;
    if (placeHolderCounter > 0) {
      for (int i = 0; i < placeHolderCounter; i++) {
        _applyTagMask(_placeholder, true);
        if (_charIndex <= _cursorPosition - 1) _charDeslocation += 1;
      }
    }

    while (_tagIndex >= 0 && _tagIndex < _tags.length) {
      _extraChar = '';
      var tag = _tags[_tagIndex];
      if (tag[_type] == _forcedChar) {
        _appendText(tag[_value]!);
        _incrementCharDeslocation(-_step);
      }
      _tagIndex += _step;
    }

    _cursorPosition =
        min(_cursorPosition + _charDeslocation, _maskedText.length);
    return _buildResultJson(_maskedText, _cursorPosition, maxLenght);
  }

  void _applyTagMask(String char, bool isPlaceHolder) {
    if (_tagIndex < 0 || _tagIndex >= _tags.length) {
      _overflow = true;
      return;
    }
    var tag = _tags[_tagIndex];
    String tagType = tag[_type]!;
    String tagValue = tag[_value]!;

    switch (tagType) {
      case _fixChar:
        _appendExtraChar(tagValue);
        _tagIndex += _step;
        _applyTagMask(char, isPlaceHolder);
        break;
      case _forcedChar:
        _appendText(tagValue);
        _incrementCharDeslocation(1);
        _tagIndex += _step;
        _applyTagMask(char, isPlaceHolder);
        break;
      case _token:
        if (_match(tagValue, char) || isPlaceHolder) {
          _appendText(char);
          _typedCharacter += 1;
          _tagIndex += _step;
        } else {
          _incrementCharDeslocation(-1);
        }
        break;
      case _tokenOpt:
        if (_match(tagValue, char) || isPlaceHolder) {
          _appendText(char);
          _typedCharacter += 1;
          _tagIndex += _step;
        } else {
          _tagIndex += _step;
          _applyTagMask(char, isPlaceHolder);
        }
        break;
      case _multiple:
        if (_match(tagValue, char) || isPlaceHolder) {
          _appendText(char);
          _typedCharacter += 1;
          tag['readed'] = '1';
        } else if (tag['readed']!.isNotEmpty) {
          _tagIndex += _step;
          _applyTagMask(char, isPlaceHolder);
        } else {
          _incrementCharDeslocation(-1);
        }
        break;
      case _multipleOpt:
        if (_match(tagValue, char) || isPlaceHolder) {
          _appendText(char);
          _typedCharacter += 1;
        } else {
          _tagIndex += _step;
          _applyTagMask(char, isPlaceHolder);
        }
        break;
      default:
        _incrementCharDeslocation(-1);
    }
  }

  void _incrementCharDeslocation(int step) {
    if (_charIndex <= _cursorPosition - 1) _charDeslocation += step;
  }

  bool _match(String tagValue, String char) => RegExp(tagValue).hasMatch(char);

  void _appendText(String char) {
    _maskedText = _reverse
        ? '$char$_extraChar$_maskedText'
        : '$_maskedText$_extraChar$char';
    _incrementCharDeslocation(_extraChar.length);
    _extraChar = '';
  }

  void _appendExtraChar(String extra) {
    _extraChar = _reverse ? '$extra$_extraChar' : '$_extraChar$extra';
  }

  Map<String, dynamic> _buildResultJson(
      String text, int cursorPos, int maxLengh) {
    if (maxLengh > 0) {
      if (_reverse) {
        text = text.substring(max(0, text.length - maxLengh));
      } else {
        text = text.substring(0, maxLengh);
      }
    }
    return <String, dynamic>{
      "text": text,
      "selectionBase": max(0, cursorPos),
      "selectionExtent": max(0, cursorPos),
      "overflow": _overflow
    };
  }
}


lib/src/text_input_mask.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'magic_mask.dart';

/// TextInputMask extends the TextInputFormatter to make your life better!
/// Just initiate it with the mask as string and it's done.
class TextInputMask extends TextInputFormatter {
  dynamic mask;
  String placeholder;
  bool reverse;
  int maxLength;
  int maxPlaceHolders;
  late MagicMask magicMask;

  /// [mask] is the String or Array of Strings to be used as mask(s).
  /// [reverse] is a bool. When true it will mask on reverse mode, usually to be used on currency fields.
  /// [maxLength] can be used to limit the maximum length. Leave it null or -1 to not limited
  /// [placeholder] is a string to be applied on untyped characters.
  /// [maxPlaceHolders] max times a placeholder is counted. Typed characters consumes the counter.
  /// ex placeholder as '0' with max=3 on a text like '3' with mask 9+.99 will be 0.03 not 000000.03
  ///
  /// The allowed patterns to it are:
  ///
  /// 9 - is used to allow a number from 0-9
  ///
  /// A - is used to allow a letter from a-z or A-Z
  ///
  /// N - is used to allow a number or letter from 0-9, a-z or A-Z
  ///
  /// X - is used to allow any character
  ///
  /// **Those tokens 9,A,N and X can be followed by one following modifier**
  ///
  /// \? - indicates that is optional
  ///
  /// \+ - indicates that must have at least 1 or more repetitions
  ///
  /// \* - indicates that can have 0 or more repetitions
  ///
  /// \ - is used as scape
  ///
  /// ** Any character that is interpreted as letter to be placed, can be followed by modifier **
  ///
  /// \! - Used to force print it, when it has at least 1 letter
  ///
  /// When passing an array of String as mask, the first mask applied is the shortest going to longest.
  /// It will apply the next mask (bigger one, only when the typed text overflow the previous mask)
  TextInputMask(
      {this.mask,
      this.reverse = false,
      this.maxLength = -1,
      this.placeholder = '',
      this.maxPlaceHolders = -1}) {
    magicMask = MagicMask.buildMask(mask);
  }

  @override
  TextEditingValue formatEditUpdate(
      TextEditingValue oldValue, TextEditingValue newValue) {
    try {
      return TextEditingValue.fromJSON(
        magicMask.executeMasking(
          newValue.text,
          newValue.selection.baseOffset,
          reverse,
          maxLength,
          placeholder,
          maxPlaceHolders,
        ),
      );
    } catch (e) {
      print(e);
    }
    return newValue;
  }
}


example/lib/main.dart
import 'package:easy_mask/easy_mask.dart';
import 'package:flutter/material.dart';

// import 'package:easy_mask.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              padding: EdgeInsets.symmetric(horizontal: 60),
              child: TextField(
                decoration: InputDecoration(hintText: 'Cellphone'),
                inputFormatters: [
                  TextInputMask(mask: '\\+ 9 (999) 9999 99 99', reverse: false)
                ],
              ),
            ),
            Container(
              padding: EdgeInsets.symmetric(horizontal: 60),
              child: TextField(
                decoration: InputDecoration(hintText: 'Brazilian CPF'),
                inputFormatters: [
                  TextInputMask(
                      mask: '999.999.999-99',
                      placeholder: '_',
                      maxPlaceHolders: 11,
                      reverse: false)
                ],
              ),
            ),
            Container(
              padding: EdgeInsets.symmetric(horizontal: 60),
              child: TextField(
                decoration: InputDecoration(hintText: 'Multi Mask CPF/CNPJ'),
                inputFormatters: [
                  TextInputMask(
                      mask: ['999.999.999-99', '99.999.999/9999-99'],
                      reverse: false)
                ],
              ),
            ),
            Container(
              padding: EdgeInsets.symmetric(horizontal: 60),
              child: TextField(
                decoration: InputDecoration(hintText: 'Multi Mask Phone'),
                inputFormatters: [
                  TextInputMask(
                      mask: ['(99) 9999 9999', '(99) 99999 9999'],
                      reverse: false)
                ],
              ),
            ),
            Container(
              padding: EdgeInsets.symmetric(horizontal: 60),
              child: TextField(
                decoration: InputDecoration(hintText: 'Date'),
                inputFormatters: [
                  TextInputMask(mask: '99/99/9999', reverse: false)
                ],
              ),
            ),
            Container(
              padding: EdgeInsets.symmetric(horizontal: 60),
              child: TextField(
                decoration: InputDecoration(hintText: 'Brazilian Phone'),
                inputFormatters: [
                  TextInputMask(mask: '\\+5!5! (!99) 99999-9999')
                ],
              ),
            ),
            Container(
              padding: EdgeInsets.symmetric(horizontal: 60),
              child: TextField(
                decoration: InputDecoration(hintText: 'Brazilian Car Plate'),
                inputFormatters: [
                  TextInputMask(mask: 'AAA9N99')
                ],
              ),
            ),
            Container(
              padding: EdgeInsets.symmetric(horizontal: 120),
              child: TextField(
                textAlign: TextAlign.right,
                decoration: InputDecoration(hintText: 'Money'),
                inputFormatters: [
                  TextInputMask(
                    mask: '\$! !9+,99',
                    placeholder: '0',
                    maxPlaceHolders: 3,
                    reverse: true,
                  )
                ],
              ),
            )
          ],
        ),
      ),
    );
  }
}
